### ISSUE TITLE
Vue watch ?

### ISSUE THREAD

**Issue by rubydesign** [rubys/ruby2js#74]
Hi, 
i'm back at using Vue, this time with three.js. And i found a useful Vue feature to propagate my UI changes into the 3d world, called watch.
Apparently one can give a watch key to the constructor (like the methods and computed), and this will watch for any data changes and trigger a given code. This is just what i need to update my 3d model with one line, and not have to do it for every attribute separately, as watch supports recursive mode.

I tried all i could to get ruby2js to output the watch, but could not. Then i looked at the code (which i don't fully understand) and remembered that computed and methods are separated automagically by whether they have args or not. So i don't really see how the "watch" could fall into that.

More digging has revealed a new Vue method Vue.$watch, that can sort of do a similar effect. So I might not need the watch in the constructor.
So i guess this is more of a question, do you want to support this feature? And would there be an "easy" way to do so?

https://vuejs.org/v2/api/#watch
https://vuejs.org/v2/api/#vm-watch

**Comment by rubys** [comment]
Hi Torsten, welcome back!

Would it be possible for you to describe what transformation you would like to see?  Don't worry too much about the current implementation, I would prefer that you describe what you would see as the ideal Ruby syntax for a watch would be, and what that JavaScript syntax should be generated from that input.

That would form the basis for test cases, and I will look into implementing the necessary filter changes.

**Comment by rubydesign** [comment]
Hi Sam, and thanks. Fun to do some more 3d

maybe the most pressing thing is this line of js, with Numbers being my Vue instance
`Numbers.$watch("_data" ,  refreshText  , {deep: true})`
a way to move that into ruby would be great, currently the $ is stopping that.

More generally, i am not sure. The `watch` key can take several kinds of values (as seen in the docs).
Maybe having a `watch` function on the Vue instance would be a way. The contents of the method would end up being transformed and used as the value for a `watch` key on the constructor. 

In code maybe
```
class Numbers < Vue
 . . . . .
  def watch
       { "_data":  refreshText  , deep: true  }
  end
end
```

Transforming to js
```
const Numbers = new Vue({el: "#info", data: ...
  watch:  { "_data" =>  refreshText  , deep: true  }
}})
```
Basically passing the hash though. It could even check that a it is a hash and no random code.
Some docs would be needed though. 

PS: code is live at https://rubydesign.fi/pages/numbers
Previous at: https://ecoframe.herokuapp.com/en/house

**Comment by rubys** [comment]
@rubydesign  I've committed support for that particular mapping, the code changes are straightforward:

https://github.com/rubys/ruby2js/compare/9f16b16c18855a5855cbf74350de5e2b01571bb3...838c093c5bc3165089503f262bdd495956f0a0b3

Does this meet your needs?  Is there more that needs to be done?

**Comment by rubydesign** [comment]
Ok, thanks. 
But when i try i get 
  `methods: {watch() {{_data: refreshText, deep: true}}}`
which in a way makes the test pass, but the watch is at the wrong level.
My initializer is 
```
require "ruby2js"
require "ruby2js/es2015"
require "ruby2js/filter/functions"
require "ruby2js/haml"
require "ruby2js/filter/vue"
require "ruby2js/rails"
```
just saying, because i wonder about the syntax change from `watch: ...` to `watch(){...}`
which i think are equivalent (in some universe that i don't inhabit)

In short, this didn't do the trick, and i don't know why

**Comment by rubys** [comment]
@rubydesign are you sure you pulled the latest code?  I've updated the test to verify the full output:

https://github.com/rubys/ruby2js/commit/b13801bcfdf718a5fbcc77aba9132b81f8fe4bc1

You can also experiment here:  http://intertwingly.net/projects/ruby2js.cgi/vue

**Comment by rubydesign** [comment]
Ok that works.
I'm pretty sure i put the git in the Gemfile before, but i can see you only changed the test. So i must have made it wrong.

Now i have an unexpected sidefect. Not 100% related, as the code works. But when i give it the method name it calls it on the instance, and i want to call a global function. So i have to pass a lambda. That's Ok, just for the record.  

But great, thanks

**Comment by rubys** [comment]
@rubydesign if you can provide a test case of what input you provided and what output you expected to be produced, I can take a look at it.

**Comment by rubydesign** [comment]
So if i was unclear: Your solution is good and works.

Vue works different than expected, but i just use a lambda instead of the method name and all is good.

**Comment by rubys** [comment]
cool.  gem pushed
